Add filename field to export

marthacryan opened the pull request:
## Description
This PR adds a "file name" input field into the export task pane:
![image](https://github.com/mito-ds/monorepo/assets/6673460/df10c0be-6e42-4664-9b40-23da017e07a9)

It displays a placeholder and is optional for the user, defaulting to "MitoExport":
![image](https://github.com/mito-ds/monorepo/assets/6673460/13acb2ec-d1c8-4c22-b06e-6e62bc8f489a)

It validates for forbidden characters in filenames:
![image](https://github.com/mito-ds/monorepo/assets/6673460/ab4972de-6fb4-4a2c-b07f-b8205631461b)

## Todo
- [x] Add validation
  - [x] Prevent form submission when field is invalid
- [ ] Add documentation
- [ ] Add comments to code
- [x] Update css to align all input fields

naterush commented:
@marthacryan the frontend tests tend to be flakier than is ideal - so if only one of them is failing (as is the case currently) -- feel free to ignore.

naterush commented on mitosheet/src/types.tsx:
```diff
@@ -825,8 +825,9 @@ export interface MitoStateUpdaters {
 
-export interface CSVExportState { exportType: 'csv' }
-export interface ExcelExportState { exportType: 'excel', sheetIndexes: number[] }
+export interface ExportState { fileName?: string, exportType: 'csv' | 'excel' }
+export interface CSVExportState extends ExportState { exportType: 'csv' }
+export interface ExcelExportState extends ExportState { exportType: 'excel', sheetIndexes: number[] }
```
These are pleasant types! We rarely use extends, but just b/c we forget about them. Nice!

marthacryan replied:
Thanks!

naterush commented on mitosheet/src/components/taskpanes/Download/DownloadTaskpane.tsx:
```diff
@@ -169,13 +199,35 @@ const DownloadTaskpane = (props: DownloadTaskpaneProps): JSX.Element => {
                             userProfile={props.userProfile}
                             sheetDataArray={props.sheetDataArray}
-                            exportState={props.uiState.exportConfiguration}
+                            exportState={props.uiState.exportConfiguration as ExcelExportState}
```
So sad TS can't type check this </3

naterush commented on mitosheet/src/components/elements/TextButton.tsx:
```diff
@@ -70,6 +70,11 @@ interface TextButtonProps {
     disabledTooltip?: string
+
+    /**
+     * Boolean to indicate whether the button is showing error. Impacts color. 
```
This doesn't appear to impact color any more than disabled. Can we rip this prop out and just pass things through disabled for now?

naterush commented on mitosheet/src/components/taskpanes/Download/DownloadTaskpane.tsx:
```diff
@@ -21,6 +22,18 @@ import DefaultEmptyTaskpane from '../DefaultTaskpane/DefaultEmptyTaskpane';
 import DefaultTaskpaneFooter from '../DefaultTaskpane/DefaultTaskpaneFooter';
 
 
+const INVALID_CHARACTERS_IN_FILENAME = [
```
Given that we define this constant, and the code that consumes it twice across both download taskpanes, how would you feel about moving this into a shared utility like `getInvalidFileNameError(fileName: string): string | undefined`?

naterush commented on mitosheet/src/components/taskpanes/Download/DownloadTaskpane.tsx:
```diff
@@ -207,13 +259,14 @@ const DownloadTaskpane = (props: DownloadTaskpaneProps): JSX.Element => {
                 >
                     
-                    {exportString === '' ? (<>Preparing data for download <LoadingDots /></>) : `Download ${props.uiState.exportConfiguration.exportType === 'csv' ? 'CSV file': 'Excel workbook'}`}
+                    {invalidFileNameWarning ? <>Invalid input.</> : exportString === '' ? (<>Preparing data for download <LoadingDots /></>) : `Download ${props.uiState.exportConfiguration.exportType === 'csv' ? 'CSV file': 'Excel workbook'}`}
```
Maybe slightly more helpful: 

```suggestion
                    {invalidFileNameWarning ? <>{invalidFileNameWarning}</> : exportString === '' ? (<>Preparing data for download <LoadingDots /></>) : `Download ${props.uiState.exportConfiguration.exportType === 'csv' ? 'CSV file': 'Excel workbook'}`}
```

marthacryan replied:
I guess what I was thinking is that if we add any more validation that we'd want it to be more general (in case there were multiple validation errors). But now that I'm thinking about it, I do think that generally when a form is invalid, the submission button becomes disabled but the actual text stays the same, and then you'd have more info in a tooltip or add some red text above the submission button. Maybe we can brainstorm a little about the design

naterush reviewed (changes requested):
This looks awesome -- great work getting this up today. Your code is clean, works well, and touches the exact bit of code that makes sense. Couldn't ask for more --  whenever I touch new codebases I find myself mucking about with random/wrong internals :) 

Two minor changes that I missed when specifying, but I think would be nice polish before merging: 

## Reordering/renaming the fields

Namely, so they match between the two download taskpanes. I think keeping the ExportToFileTaskpane as is, and then:
1. changing the order of file name and file type, 
2. renaming from export type to file type in this taskpane 

would be a nice easy way to standardize. 

## Button flickering

The `URL.createObjectURL(new Blob(` code reruns whenever you edit the file name, which leads to a weird flickering (which is worse with larger files).

Probably the easiest thing to do would be to store the exportHRef with a `useState`, and then just update it when you load a new exportString (might even be able to stop storing exportString). There are probably other ways of handling this with a useEffect or something -- so feel free to take whatever approach makes the most sense for you. 

Let me know thoughts on these, if you think other approaches might be better! Sorry for missing em in the spec :}

marthacryan commented:
> 1. changing the order of file name and file type,
> 2. renaming from export type to file type in this taskpane
> 
> would be a nice easy way to standardize.

Updated!

> Probably the easiest thing to do would be to store the exportHRef with a `useState`, and then just update it when you load a new exportString (might even be able to stop storing exportString). There are probably other ways of handling this with a useEffect or something -- so feel free to take whatever approach makes the most sense for you.

Updated the code to just use `exportHref` as a state variable instead of `exportString` (because `exportString` was only being used by `exportHref`). Let me know how that looks!

naterush commented:
Flickering is reduced! It's still sometimes here, I think because the useDebouncedEffect has the entire export state as it's dependency (which includes the file name). Luckily, the debouncing reducing the impact of this.

LGTM! I'll squash those merge conflicts :-)
